commit.go

```diff
@@ -1,310 +1,199 @@
+		com, err := parsePersonInfo(line)
+		if err != nil {
+			return err
 		}
 
-		return &node.Link{Cid: c.Parents[i]}, path[2:], nil
-	case "author":
-		if len(path) == 1 {
-			return c.Author, nil, nil
-		}
-		return c.Author.resolve(path[1:])
-	case "committer":
-		if len(path) == 1 {
-			return c.Committer, nil, nil
-		}
-		return c.Committer.resolve(path[1:])
-	case "signature":
-		return c.Sig.Text, path[1:], nil
-	case "message":
-		return c.Message, path[1:], nil
-	case "tree":
-		return &node.Link{Cid: c.GitTree}, path[1:], nil
-	case "mergetag":
-		if len(path) == 1 {
-			return c.MergeTag, nil, nil
+		c.Committer = _PersonInfo__Maybe{m: schema.Maybe_Value, v: com}
+	case bytes.HasPrefix(line, []byte("encoding ")):
+		c.Encoding = _String__Maybe{m: schema.Maybe_Value, v: &_String{string(line[9:])}}
+	case bytes.HasPrefix(line, []byte("mergetag object ")):
+		sha, err := hex.DecodeString(string(line)[16:])
+		if err != nil {
+			return err
 		}
 
-		i, err := strconv.Atoi(path[1])
+		mt, rest, err := readMergeTag(sha, rd)
 		if err != nil {
-			return nil, nil, err
+			return err
 		}
 
-		if i < 0 || i >= len(c.MergeTag) {
-			return nil, nil, fmt.Errorf("index out of range")
-		}
+		c.MergeTag.x = append(c.MergeTag.x, *mt)
 
-		if len(path) == 2 {
-			return c.MergeTag[i], nil, nil
+		if rest != nil {
+			err = decodeCommitLine(c, rest, rd)
+			if err != nil {
+				return err
+			}
+		}
+	case bytes.HasPrefix(line, []byte("gpgsig ")):
+		sig, err := decodeGpgSig(rd)
+		if err != nil {
+			return err
+		}
+		c.Sig = _GpgSig__Maybe{m: schema.Maybe_Value, v: sig}
+	case len(line) == 0:
+		rest, err := ioutil.ReadAll(rd)
+		if err != nil {
+			return err
 		}
-		return c.MergeTag[i].resolve(path[2:])
+
+		c.Message = _String{string(rest)}
 	default:
-		return nil, nil, errors.New("no such link")
+		c.Other.x = append(c.Other.x, _String{string(line)})
 	}
+	return nil
 }
 
-func (c *Commit) ResolveLink(path []string) (*node.Link, []string, error) {
-	out, rest, err := c.Resolve(path)
+func decodeGpgSig(rd *bufio.Reader) (GpgSig, error) {
+	line, _, err := rd.ReadLine()
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	lnk, ok := out.(*node.Link)
-	if !ok {
-		return nil, nil, errors.New("not a link")
+	out := _GpgSig{}
+
+	if string(line) != " " {
+		if strings.HasPrefix(string(line), " Version: ") || strings.HasPrefix(string(line), " Comment: ") {
+			out.x += string(line) + "\n"
+		} else {
+			return nil, fmt.Errorf("expected first line of sig to be a single space or version")
+		}
+	} else {
+		out.x += " \n"
 	}
 
-	return lnk, rest, nil
-}
+	for {
+		line, _, err := rd.ReadLine()
+		if err != nil {
+			return nil, err
+		}
 
-func (c *Commit) Size() (uint64, error) {
-	return uint64(len(c.RawData())), nil
-}
+		if bytes.Equal(line, []byte(" -----END PGP SIGNATURE-----")) {
+			break
+		}
 
-func (c *Commit) Stat() (*node.NodeStat, error) {
-	return &node.NodeStat{}, nil
-}
+		out.x += string(line) + "\n"
```

Comment:
I'd strongly advise you to use something like strings.Builder instead; each `+=` on a string is an allocation and copy of the entire thing. Just twenty of these operations could already add up to showing up on CPU profiles.

Reply:
this is copied verbatim from https://github.com/ipfs/go-ipld-git/blob/master/git.go#L310

I'll file as an issue to optimize if we end up caring